File: backend/pkg/errortracking/sentry.go
package errortracking

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"runtime"
	"time"

	"github.com/getsentry/sentry-go"
)

// Config holds the configuration for error tracking
type Config struct {
	DSN              string
	Environment      string
	Release          string
	SampleRate       float64
	TracesSampleRate float64
	Debug            bool
	ServerName       string
	BeforeSend       func(event *sentry.Event, hint *sentry.EventHint) *sentry.Event
}

// DefaultConfig returns a default configuration
func DefaultConfig() Config {
	return Config{
		DSN:              os.Getenv("SENTRY_DSN"),
		Environment:      getEnvOrDefault("ENVIRONMENT", "development"),
		Release:          getEnvOrDefault("APP_VERSION", "unknown"),
		SampleRate:       1.0,
		TracesSampleRate: 0.1,
		Debug:            os.Getenv("SENTRY_DEBUG") == "true",
		ServerName:       getHostname(),
	}
}

// Init initializes the Sentry SDK with the provided configuration
func Init(cfg Config) error {
	if cfg.DSN == "" {
		return fmt.Errorf("sentry DSN is required")
	}

	err := sentry.Init(sentry.ClientOptions{
		Dsn:              cfg.DSN,
		Environment:      cfg.Environment,
		Release:          cfg.Release,
		SampleRate:       cfg.SampleRate,
		TracesSampleRate: cfg.TracesSampleRate,
		Debug:            cfg.Debug,
		ServerName:       cfg.ServerName,
		BeforeSend:       cfg.BeforeSend,
		AttachStacktrace: true,
		EnableTracing:    true,
		Integrations: func(integrations []sentry.Integration) []sentry.Integration {
			// Filter out integrations if needed
			return integrations
		},
	})
	if err != nil {
		return fmt.Errorf("failed to initialize sentry: %w", err)
	}

	return nil
}

// InitWithDefaults initializes Sentry with default configuration
func InitWithDefaults() error {
	cfg := DefaultConfig()
	cfg.BeforeSend = DefaultBeforeSend
	return Init(cfg)
}

// DefaultBeforeSend is the default event filter
func DefaultBeforeSend(event *sentry.Event, hint *sentry.EventHint) *sentry.Event {
	// Filter out sensitive data
	event = sanitizeEvent(event)

	// Filter out specific errors
	if shouldFilterError(event, hint) {
		return nil
	}

	return event
}

// Flush waits for pending events to be sent
func Flush(timeout time.Duration) bool {
	return sentry.Flush(timeout)
}

// Close flushes and closes the Sentry client
func Close() {
	sentry.Flush(2 * time.Second)
}

// CaptureException captures an error with optional context
func CaptureException(err error, ctx ...context.Context) *sentry.EventID {
	if err == nil {
		return nil
	}

	hub := getHub(ctx...)
	return hub.CaptureException(err)
}

// CaptureMessage captures a message with optional context
func CaptureMessage(message string, ctx ...context.Context) *sentry.EventID {
	hub := getHub(ctx...)
	return hub.CaptureMessage(message)
}

// CaptureError captures an error with additional context and tags
func CaptureError(err error, opts ...ErrorOption) *sentry.EventID {
	if err == nil {
		return nil
	}

	options := &errorOptions{}
	for _, opt := range opts {
		opt(options)
	}

	hub := sentry.CurrentHub().Clone()

	// Add tags
	for key, value := range options.tags {
		hub.Scope().SetTag(key, value)
	}

	// Add extra context
	for key, value := range options.extra {
		hub.Scope().SetExtra(key, value)
	}

	// Set user if provided
	if options.user != nil {
		hub.Scope().SetUser(*options.user)
	}

	// Set level if provided
	if options.level != "" {
		hub.Scope().SetLevel(options.level)
	}

	// Add fingerprint for grouping
	if len(options.fingerprint) > 0 {
		hub.Scope().SetFingerprint(options.fingerprint)
	}

	return hub.CaptureException(err)
}

// ErrorOption is a functional option for error capture
type ErrorOption func(*errorOptions)

type errorOptions struct {
	tags        map[string]string
	extra       map[string]interface{}
	user        *sentry.User
	level       sentry.Level
	fingerprint []string
	context     context.Context
}

// WithTags adds tags to the error
func WithTags(tags map[string]string) ErrorOption {
	return func(o *errorOptions) {
		if o.tags == nil {
			o.tags = make(map[string]string)
		}
		for k, v := range tags {
			o.tags[k] = v
		}
	}
}

// WithTag adds a single tag to the error
func WithTag(key, value string) ErrorOption {
	return func(o *errorOptions) {
		if o.tags == nil {
			o.tags = make(map[string]string)
		}
		o.tags[key] = value
	}
}

// WithExtra adds extra context to the error
func WithExtra(extra map[string]interface{}) ErrorOption {
	return func(o *errorOptions) {
		if o.extra == nil {
			o.extra = make(map[string]interface{})
		}
		for k, v := range extra {
			o.extra[k] = v
		}
	}
}

// WithContext adds a single context value
func WithContext(key string, value interface{}) ErrorOption {
	return func(o *errorOptions) {
		if o.extra == nil {
			o.extra = make(map[string]interface{})
		}
		o.extra[key] = value
	}
}

// WithUser adds user information to the error
func WithUser(id, email, username string) ErrorOption {
	return func(o *errorOptions) {
		o.user = &sentry.User{
			ID:       id,
			Email:    email,
			Username: username,
		}
	}
}

// WithLevel sets the error level
func WithLevel(level sentry.Level) ErrorOption {
	return func(o *errorOptions) {
		o.level = level
	}
}

// WithFingerprint sets custom fingerprint for error grouping
func WithFingerprint(fingerprint ...string) ErrorOption {
	return func(o *errorOptions) {
		o.fingerprint = fingerprint
	}
}

// WithRequestContext adds HTTP request context
func WithRequestContext(r *http.Request) ErrorOption {
	return func(o *errorOptions) {
		if o.extra == nil {
			o.extra = make(map[string]interface{})
		}
		o.extra["request"] = map[string]interface{}{
			"method":     r.Method,
			"url":        r.URL.String(),
			"user_agent": r.UserAgent(),
			"remote_ip":  getClientIP(r),
		}
	}
}

// StartSpan starts a new Sentry span for tracing
func StartSpan(ctx context.Context, operation string, opts ...sentry.SpanOption) *sentry.Span {
	return sentry.StartSpan(ctx, operation, opts...)
}

// StartTransaction starts a new Sentry transaction
func StartTransaction(ctx context.Context, name, operation string) *sentry.Span {
	return sentry.StartTransaction(ctx, name, sentry.WithOpName(operation))
}

// RecoverWithContext recovers from panics and reports to Sentry
func RecoverWithContext(ctx context.Context) {
	if err := recover(); err != nil {
		hub := sentry.GetHubFromContext(ctx)
		if hub == nil {
			hub = sentry.CurrentHub().Clone()
		}

		// Capture stack trace
		buf := make([]byte, 64<<10)
		buf = buf[:runtime.Stack(buf, false)]

		hub.Scope().SetExtra("stacktrace", string(buf))

		switch v := err.(type) {
		case error:
			hub.CaptureException(v)
		case string:
			hub.CaptureMessage(v)
		default:
			hub.CaptureMessage(fmt.Sprintf("%v", v))
		}

		hub.Flush(2 * time.Second)

		// Re-panic after reporting
		panic(err)
	}
}

// Helper functions

func getHub(ctx ...context.Context) *sentry.Hub {
	if len(ctx) > 0 && ctx[0] != nil {
		if hub := sentry.GetHubFromContext(ctx[0]); hub != nil {
			return hub
		}
	}
	return sentry.CurrentHub()
}

func getEnvOrDefault(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

func getHostname() string {
	hostname, err := os.Hostname()
	if err != nil {
		return "unknown"
	}
	return hostname
}

func getClientIP(r *http.Request) string {
	// Check X-Forwarded-For header
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		return xff
	}
	// Check X-Real-IP header
	if xri := r.Header.Get("X-Real-IP"); xri != "" {
		return xri
	}
	return r.RemoteAddr
}

func sanitizeEvent(event *sentry.Event) *sentry.Event {
	// Remove sensitive headers
	sensitiveHeaders := []string{
		"Authorization",
		"Cookie",
		"Set-Cookie",
		"X-API-Key",
		"X-Auth-Token",
	}

	if event.Request != nil {
		for _, header := range sensitiveHeaders {
			delete(event.Request.Headers, header)
		}
	}

	// Redact sensitive extra data
	sensitiveKeys := []string{
		"password",
		"secret",
		"token",
		"api_key",
		"apikey",
		"credit_card",
		"ssn",
	}

	for key := range event.Extra {
		for _, sensitive := range sensitiveKeys {
			if containsInsensitive(key, sensitive) {
				event.Extra[key] = "[REDACTED]"
			}
		}
	}

	return event
}

func shouldFilterError(event *sentry.Event, hint *sentry.EventHint) bool {
	// Filter out specific error types
	if hint != nil && hint.OriginalException != nil {
		errMsg := hint.OriginalException.Error()

		// Filter context canceled errors
		if errMsg == "context canceled" {
			return true
		}

		// Filter client disconnection errors
		if containsInsensitive(errMsg, "client disconnected") ||
			containsInsensitive(errMsg, "broken pipe") ||
			containsInsensitive(errMsg, "connection reset by peer") {
			return true
		}
	}

	return false
}

func containsInsensitive(s, substr string) bool {
	return len(s) >= len(substr) &&
		(s == substr ||
			len(s) > len(substr) &&
				(containsLower(s, substr)))
}

func containsLower(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if equalFoldAt(s, i, substr) {
			return true
		}
	}
	return false
}

func equalFoldAt(s string, i int, substr string) bool {
	for j := 0; j < len(substr); j++ {
		c1 := s[i+j]
		c2 := substr[j]
		if c1 != c2 && toLower(c1) != toLower(c2) {
			return false
		}
	}
	return true
}

func toLower(c byte) byte {
	if c >= 'A' && c <= 'Z' {
		return c + 32
	}
	return c
}

File: frontend/src/services/errorTracking/sentry.ts
import * as Sentry from '@sentry/react';

// Error tracking configuration
export interface ErrorTrackingConfig {
  dsn: string;
  environment: string;
  release?: string;
  sampleRate?: number;
  tracesSampleRate?: number;
  debug?: boolean;
  allowUrls?: (string | RegExp)[];
  denyUrls?: (string | RegExp)[];
  beforeSend?: (
    event: Sentry.Event,
    hint: Sentry.EventHint
  ) => Sentry.Event | null;
}

// Default configuration
const defaultConfig: Partial<ErrorTrackingConfig> = {
  environment: import.meta.env.MODE || 'development',
  release: import.meta.env.VITE_APP_VERSION || 'unknown',
  sampleRate: 1.0,
  tracesSampleRate: 0.1,
  debug: import.meta.env.DEV,
};

// Sensitive data patterns to filter
const sensitivePatterns = [
  /password/i,
  /secret/i,
  /token/i,
  /api_key/i,
  /apikey/i,
  /authorization/i,
  /credit_card/i,
  /creditcard/i,
  /ssn/i,
  /social_security/i,
];

// URLs to ignore (usually third-party scripts)
const defaultDenyUrls = [
  /extensions\//i,
  /^chrome:\/\//i,
  /^chrome-extension:\/\//i,
  /^moz-extension:\/\//i,
];

// Error messages to filter out
const ignoredErrors = [
  'ResizeObserver loop limit exceeded',
  'ResizeObserver loop completed with undelivered notifications',
  'Non-Error promise rejection captured',
  'Network Error',
  'Load failed',
  'Failed to fetch',
  'ChunkLoadError',
  'Loading chunk',
  'cancelled',
  'AbortError',
];

/**
 * Initialize Sentry error tracking
 */
export function initErrorTracking(config: ErrorTrackingConfig): void {
  if (!config.dsn) {
    console.warn('Sentry DSN not configured. Error tracking disabled.');
    return;
  }

  const finalConfig = { ...defaultConfig, ...config };

  Sentry.init({
    dsn: finalConfig.dsn,
    environment: finalConfig.environment,
    release: finalConfig.release,
    sampleRate: finalConfig.sampleRate,
    tracesSampleRate: finalConfig.tracesSampleRate,
    debug: finalConfig.debug,

    integrations: [
      Sentry.browserTracingIntegration({
        enableInp: true,
      }),
      Sentry.replayIntegration({
        maskAllText: true,
        blockAllMedia: true,
      }),
    ],

    tracePropagationTargets: [
      'localhost',
      /^https:\/\/api\.servicepro\.com/,
      /^https:\/\/.*\.servicepro\.com/,
    ],

    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,

    allowUrls: finalConfig.allowUrls,
    denyUrls: [...defaultDenyUrls, ...(finalConfig.denyUrls || [])],

    beforeSend: (event, hint) => {
      // Apply custom filter first
      if (finalConfig.beforeSend) {
        const result = finalConfig.beforeSend(event, hint);
        if (!result) return null;
        // Use the filtered result
        return sanitizeEvent(result) as Sentry.ErrorEvent;
      }

      // Filter ignored errors
      if (shouldIgnoreError(event, hint)) {
        return null;
      }

      // Sanitize sensitive data
      return sanitizeEvent(event) as Sentry.ErrorEvent;
    },

    beforeBreadcrumb: (breadcrumb) => {
      // Filter out noisy breadcrumbs
      if (breadcrumb.category === 'console' && breadcrumb.level === 'debug') {
        return null;
      }

      // Sanitize breadcrumb data
      if (breadcrumb.data) {
        breadcrumb.data = sanitizeData(breadcrumb.data);
      }

      return breadcrumb;
    },
  });
}

/**
 * Check if an error should be ignored
 */
function shouldIgnoreError(
  event: Sentry.Event,
  hint: Sentry.EventHint
): boolean {
  const error = hint.originalException;

  // Check error message
  if (error instanceof Error) {
    for (const ignored of ignoredErrors) {
      if (error.message.includes(ignored)) {
        return true;
      }
    }
  }

  // Check event message
  if (event.message) {
    for (const ignored of ignoredErrors) {
      if (event.message.includes(ignored)) {
        return true;
      }
    }
  }

  return false;
}

/**
 * Sanitize event data to remove sensitive information
 */
function sanitizeEvent(event: Sentry.Event): Sentry.Event {
  // Sanitize request headers
  if (event.request?.headers) {
    const sensitiveHeaders = [
      'Authorization',
      'Cookie',
      'Set-Cookie',
      'X-API-Key',
    ];
    for (const header of sensitiveHeaders) {
      if (event.request.headers[header]) {
        event.request.headers[header] = '[FILTERED]';
      }
    }
  }

  // Sanitize extra data
  if (event.extra) {
    event.extra = sanitizeData(event.extra);
  }

  // Sanitize tags
  if (event.tags) {
    event.tags = sanitizeData(event.tags) as Record<string, string>;
  }

  return event;
}

/**
 * Sanitize object data recursively
 */
function sanitizeData(data: Record<string, unknown>): Record<string, unknown> {
  const sanitized: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(data)) {
    // Check if key matches sensitive pattern
    const isSensitive = sensitivePatterns.some((pattern) => pattern.test(key));

    if (isSensitive) {
      sanitized[key] = '[REDACTED]';
    } else if (typeof value === 'object' && value !== null) {
      sanitized[key] = sanitizeData(value as Record<string, unknown>);
    } else {
      sanitized[key] = value;
    }
  }

  return sanitized;
}

/**
 * Capture an exception with additional context
 */
export function captureException(
  error: Error | unknown,
  context?: {
    tags?: Record<string, string>;
    extra?: Record<string, unknown>;
    user?: { id: string; email?: string; username?: string };
    level?: Sentry.SeverityLevel;
    fingerprint?: string[];
  }
): string | undefined {
  return Sentry.withScope((scope) => {
    if (context?.tags) {
      for (const [key, value] of Object.entries(context.tags)) {
        scope.setTag(key, value);
      }
    }

    if (context?.extra) {
      for (const [key, value] of Object.entries(context.extra)) {
        scope.setExtra(key, value);
      }
    }

    if (context?.user) {
      scope.setUser(context.user);
    }

    if (context?.level) {
      scope.setLevel(context.level);
    }

    if (context?.fingerprint) {
      scope.setFingerprint(context.fingerprint);
    }

    return Sentry.captureException(error);
  });
}

/**
 * Capture a message
 */
export function captureMessage(
  message: string,
  level: Sentry.SeverityLevel = 'info',
  context?: Record<string, unknown>
): string | undefined {
  return Sentry.withScope((scope) => {
    scope.setLevel(level);

    if (context) {
      for (const [key, value] of Object.entries(context)) {
        scope.setExtra(key, value);
      }
    }

    return Sentry.captureMessage(message);
  });
}

/**
 * Set user context for error tracking
 */
export function setUser(
  user: { id: string; email?: string; username?: string } | null
): void {
  Sentry.setUser(user);
}

/**
 * Set a tag on all future events
 */
export function setTag(key: string, value: string): void {
  Sentry.setTag(key, value);
}

/**
 * Set extra context on all future events
 */
export function setExtra(key: string, value: unknown): void {
  Sentry.setExtra(key, value);
}

/**
 * Add a breadcrumb
 */
export function addBreadcrumb(breadcrumb: Sentry.Breadcrumb): void {
  Sentry.addBreadcrumb(breadcrumb);
}

/**
 * Start a performance span (replaces deprecated startTransaction)
 */
export function startPerformanceSpan<T>(
  name: string,
  op: string,
  callback: () => T
): T {
  return Sentry.startSpan({ name, op }, callback);
}

/**
 * Get the active span
 */
export function getActiveSpan(): Sentry.Span | undefined {
  return Sentry.getActiveSpan();
}

/**
 * Flush pending events
 */
export async function flush(timeout = 2000): Promise<boolean> {
  return Sentry.flush(timeout);
}

/**
 * Close the Sentry client
 */
export async function close(timeout = 2000): Promise<boolean> {
  return Sentry.close(timeout);
}

// Export Sentry for direct access when needed
export { Sentry };
